src/se/vidstige/jadb/JadbConnection.java

```diff
@@ -38,9 +39,35 @@ public void getHostVersion() throws IOException, JadbException {
         devices.send("host:devices");
         devices.verifyResponse();
         String body = devices.readString();
+        devices.close();
         return parseDevices(body);
     }
 
+    public AsyncActionHandler watchDevices(final DeviceDetectionListener listener) throws IOException, JadbException {
+        final Transport transport = createTransport();
+        new Thread(new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    transport.send("host:track-devices");
+                    transport.verifyResponse();
+                    boolean r = false;
+                    do {
+                        List<JadbDevice> list = parseDevices(transport.readString());
+                        r = listener.detect(list);
+                    } while (r);
+                } catch (SocketException e) {
+                    // socket closed from another thread
+                } catch (Exception e) {
+                    Thread t = Thread.currentThread();
+                    t.getUncaughtExceptionHandler().uncaughtException(t, e);
+                }
+            }
+        }).start();
+
+        return new AsyncActionHandler(transport);
+    }
```

Comment:
Well, one way is to return an object, like `DeviceWatcher` that can have a `stop` function that calls close on the socket. How about that?